// @nativescript/angular/directives/index.js
Object.defineProperty(exports, "__esModule", { value: true });
var list_view_comp_1 = require("./list-view-comp");
exports.ListViewComponent = list_view_comp_1.ListViewComponent;
var templated_items_comp_1 = require("./templated-items-comp");
exports.TemplateKeyDirective = templated_items_comp_1.TemplateKeyDirective;
exports.TemplatedItemsComponent = templated_items_comp_1.TemplatedItemsComponent;
var tab_view_1 = require("./tab-view");
exports.TabViewDirective = tab_view_1.TabViewDirective;
exports.TabViewItemDirective = tab_view_1.TabViewItemDirective;
var action_bar_1 = require("./action-bar");
exports.ActionBarComponent = action_bar_1.ActionBarComponent;
exports.ActionBarScope = action_bar_1.ActionBarScope;
exports.ActionItemDirective = action_bar_1.ActionItemDirective;
exports.NavigationButtonDirective = action_bar_1.NavigationButtonDirective;
var platform_filters_1 = require("./platform-filters");
exports.AndroidFilterComponent = platform_filters_1.AndroidFilterComponent;
exports.IosFilterComponent = platform_filters_1.IosFilterComponent;
var dialogs_1 = require("./dialogs");
exports.ModalDialogHost = dialogs_1.ModalDialogHost;
exports.ModalDialogParams = dialogs_1.ModalDialogParams;
exports.ModalDialogService = dialogs_1.ModalDialogService;
exports.NS_DIRECTIVES = [
    list_view_comp_1.ListViewComponent,
    templated_items_comp_1.TemplateKeyDirective,
    tab_view_1.TabViewDirective,
    tab_view_1.TabViewItemDirective,
    action_bar_1.ActionBarComponent,
    action_bar_1.ActionBarScope,
    action_bar_1.ActionItemDirective,
    action_bar_1.NavigationButtonDirective,
    platform_filters_1.AndroidFilterComponent,
    platform_filters_1.IosFilterComponent,
];
